In [1]:
import pandas as pd

In [2]:
data = pd.read_csv("../data/online_shoppers.csv")

In [3]:
data.head()

,Administrative,Administrative_Duration,Informational,Informational_Duration,ProductRelated,ProductRelated_Duration,BounceRates,ExitRates,PageValues,SpecialDay,Month,OperatingSystems,Browser,Region,TrafficType,VisitorType,Weekend,Revenue
0,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,1,1,1,1,Returning_Visitor,False,False
1,0,0.0,0,0.0,2,64.000000,0.00,0.10,0.0,0.0,Feb,2,2,1,2,Returning_Visitor,False,False
2,0,0.0,0,0.0,1,0.000000,0.20,0.20,0.0,0.0,Feb,4,1,9,3,Returning_Visitor,False,False
3,0,0.0,0,0.0,2,2.666667,0.05,0.14,0.0,0.0,Feb,3,2,2,4,Returning_Visitor,False,False
4,0,0.0,0,0.0,10,627.500000,0.02,0.05,0.0,0.0,Feb,3,3,1,4,Returning_Visitor,True,False


In [4]:
df = data.drop_duplicates()
df.shape

(12205, 18)

In [5]:
y = df['Revenue']
X= df.drop(columns = 'Revenue')


In [6]:
from sklearn.model_selection import train_test_split

In [7]:

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42, stratify= y)

In [8]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((9764, 17), (2441, 17), (9764,), (2441,))

In [9]:
y_train.mean()

np.float64(0.15628840639082342)

In [10]:
y_test.mean()

np.float64(0.15649324047521507)

## Split
total size after duplicated drop = (12205,18)
test rate = 0.2
Purchase rate: train 15.6%, test 15.6% (stratified)

In [11]:
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score, recall_score

In [12]:
dummy = DummyClassifier(strategy = 'most_frequent')
dummy.fit(X_train,y_train)
predictions = dummy.predict(X_train)

print(f" Accuracy : {accuracy_score(y_train, predictions):.3f} \n Recall : {recall_score(y_train,predictions)}")


 Accuracy : 0.844 
 Recall : 0.0


## Dummy baseline
- The dummy (most_frequent) always predicts majority class for every session.
- Accuracy = 84.4% because 84.4% of sessions are non-buyers, so "always no" is right that often.
- Recall = 0.0% because it never predicts minority class, so it catches none of the real buyers.
- Conclusion: accuracy is misleading on imbalanced data; we use PR-AUC instead.